### ISSUE TITLE
The docker image has a too old LLVM version

### ISSUE THREAD

**Issue by andreafioraldi** [google/fuzzbench#46]
Hi,
I'm trying to build AFL++ w/o trace-pc using LLVM 9.
However, you base image is Ubuntu 16 that is too old and has a legacy libc++.
Also installing LLVM 9 from the LLVM repos, I cannot replace libc++ w/o recompiling it (and it is not a good idea, it heavily depends in libc).

I tried 

```
ARG parent_image=gcr.io/fuzzbench/base-builder
FROM $parent_image

# Install LLVM 9
RUN apt-get update && \
    apt-get install software-properties-common -y && \
    apt-get install wget -y && \
    wget https://apt.llvm.org/llvm.sh && \
    chmod +x llvm.sh && \
    ./llvm.sh 9 && \
    apt-get install libc++-dev -y && \
    update-alternatives --install /usr/bin/cc cc /usr/bin/clang-9 100 && \
    update-alternatives --install /usr/bin/c++ c++ /usr/bin/clang++-9 100 && \
    update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-9 100 && \
    update-alternatives --install /usr/bin/clang clang /usr/bin/clang-9 100
```

But nothing works. Can you provide a base-builder with a more recent OS?
I understand that some fuzzers need Ubuntu 16 like QSYM, but LLVM 3.8 is really too old.

**Comment by vanhauser-thc** [comment]
Yes modern fuzzers should have a modern Ubuntu. How else can be use features from new llvm versions that make fuzzing better? That would make a very bad comparison ...

**Comment by jonathanmetzman** [comment]
> Hi,
> I'm trying to build AFL++ w/o trace-pc using LLVM 9.
> However, you base image is Ubuntu 16 that is too old and has a legacy libc++.
> Also installing LLVM 9 from the LLVM repos, I cannot replace libc++ w/o recompiling it (and it is not a good idea, it heavily depends in libc).
> 
> I tried
> 
> ```
> ARG parent_image=gcr.io/fuzzbench/base-builder
> FROM $parent_image
> 
> # Install LLVM 9
> RUN apt-get update && \
>     apt-get install software-properties-common -y && \
>     apt-get install wget -y && \
>     wget https://apt.llvm.org/llvm.sh && \
>     chmod +x llvm.sh && \
>     ./llvm.sh 9 && \
>     apt-get install libc++-dev -y && \
>     update-alternatives --install /usr/bin/cc cc /usr/bin/clang-9 100 && \
>     update-alternatives --install /usr/bin/c++ c++ /usr/bin/clang++-9 100 && \
>     update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-9 100 && \
>     update-alternatives --install /usr/bin/clang clang /usr/bin/clang-9 100
> ```
> 
> But nothing works. Can you provide a base-builder with a more recent OS?
> I understand that some fuzzers need Ubuntu 16 like QSYM, but LLVM 3.8 is really too old.

The builder images [install clang from OSS-Fuzz](https://github.com/google/fuzzbench/blob/master/docker/base-builder/Dockerfile#L47), which is generally a few weeks behind master (i.e. clang/llvm 10). I'll look into installing LLVM from there in a day or so. Would that solve this issue?

>I understand that some fuzzers need Ubuntu 16 like QSYM, but LLVM 3.8 is really too old.

>Yes modern fuzzers should have a modern Ubuntu. How else can be use features from new llvm versions that make fuzzing better? That would make a very bad comparison ...

I agree that ideally we would be using a newer ubuntu.
I can look into this, but I'm not optimistic we will switch soon.
One reason is because of QSYM as @andreafioraldi points out.
But a much bigger reason is because OSS-Fuzz, which is a source of benchmarks, uses this version of Ubuntu (since MSAN wont work in newer ones).
If we could still plop in many OSS-Fuzz benchmarks (like we do [here](https://github.com/google/fuzzbench/blob/master/benchmarks/bloaty_fuzz_target/oss-fuzz.yaml)), then using a newer version might be an option.

Thanks for this feedback, It definitely helps in weighing the tradeoffs here.

**Comment by jonathanmetzman** [comment]
There were two issues with using llvm_mode that I fixed and one that I haven't figured out. 
1. llvm-tools wasn't installed with clang. This was easy to do and I am going to land this fix in #61 
2. llvm_mode needs libstdc++ to compile. The images by default don't come with libstdc++, they come with libc++ instead since OSS-Fuzz does the same I figured things would be more likely to "just work" if we used the same environment. I worked around this by installing libstdc++ [here](https://github.com/google/fuzzbench/pull/61/commits/42ad327e5d0c8299613a35b3bb47b2b36bfa2918#diff-3f8ebc431c7cfb22f78325251e391f53R27) in my patch (can probably apt install it directly instead of implicitly through g++).

The last issue I haven't figured out yet which is why afl-clang-fast++ fails at building lcms.
It fails with this error message:
```
/bin/bash ../../libtool  --tag=CC   --mode=link /afl/afl-clang-fast  -O2 -fno-omit-frame-pointer -gline-tables-only -fsanitize=address -fvisibility=hidden    -o transicc transicc.o ../common/xgetopt.o ../common/vprf.o ../../src/liblcms2.la  
libtool: link: /afl/afl-clang-fast -O2 -fno-omit-frame-pointer -gline-tables-only -fsanitize=address -fvisibility=hidden -o .libs/transicc transicc.o ../common/xgetopt.o ../common/vprf.o  ../../src/.libs/liblcms2.so -lm -lpthread
transicc.o: In function `HandleSwitches':
```
One difference I noticed from compiling with afl-clang-fast that may be responsible is in [this line ](https://github.com/google/fuzzbench/runs/485997663#step:4:16593). When compiling with clang it says the visibility attribute works but not with afl-clang-fast.

Any ideas what could be wrong here?

**Comment by jonathanmetzman** [comment]
> One difference I noticed from compiling with afl-clang-fast that may be responsible is in [this line ](https://github.com/google/fuzzbench/runs/485997663#step:4:16593). When compiling with clang it says the visibility attribute works but not with afl-clang-fast.
> 
> Any ideas what could be wrong here?

This doesn't happen when compiling lcms locally with afl-clang-fast, probably something weird on the image, looking into it.

**Comment by jonathanmetzman** [comment]
The actual cause of the issue with lcms was afl-clang-fast++ writing to stderr and a silly autoconf test that considered the visibility attribute unsupported if the compiler wrote anything to stderr.

**Comment by jonathanmetzman** [comment]
I think I fixed this in #69 and #61.
